### ISSUE TITLE
Memory leaks

### ISSUE THREAD

**Issue by michael-grunder** [phpredis/phpredis#1703]
While investigating #1701 it became apparent that PHP wasn't always reporting memory leaks when running the phpredis unit tests, even when PHP was built in debug mode.

When I pulled the logic out of the unit test framework the leaks were again reported, and it was also possible to detect them with `USE_ZEND_ALLOC=0 valgrind php script-that-leaks.php`.

I'm just creating this issue as a central place to discuss leaks/fixes.

**Comment by michael-grunder** [comment]
[Leak branch here](https://github.com/phpredis/phpredis/tree/1703.memory-leaks)

**Comment by yatsukhnenko** [comment]
@michael-grunder have we already merged changes from your branch?

**Comment by michael-grunder** [comment]
No.  The leaks are actually non-trivial to fix.  I did get most of them in this branch though.

The good news is that they're pretty obscure/rare and don't likely occur very often.

I can revisit this to see about fixing the last ones I didn't get to yet.

**Comment by rqc** [comment]
Hello, first of all thanks for the work on this library, it has really helped us scale our php microservices.

I wanted to report that we are having some issues with memory leaks after integrating in a high traffic production server. 

We performed some simple debugging using `memory_get_usage()` and the delta was several megabytes more than the actual payload that was being processed in the request. We have narrowed down the leak to the usage on a rate limiting function that uses a query like this:

```php
# This is the php-redis client, we are using version 5.3.2
$this->client->cache
                ->multi(Redis::PIPELINE)
                ->hset(self::$BASE_RATE_LIMIT_COUNTER_KEY, $tokens_bucket_key, $this->token_allowance)
                ->hset(self::$BASE_RATE_LIMIT_COUNTER_KEY, $last_reset_key, $now->getTimestamp())
                ->exec();
```

We are currently using version:

```
=========================================
Package Version State
redis   5.3.2   stable
```

It would be great if you could perhaps share if there is a known leak in any of those commands or the use of pipelining. 

Thanks!
